compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp

```diff
@@ -578,6 +578,172 @@ static int64_t getMaxVectorTileSize(int64_t numElem, int64_t tileSize,
+// `allOperandsMaxTileBits`.
+static void limitVectorTileSizes(linalg::LinalgOp op,
+                                 SmallVectorImpl<int64_t> &vecTileSizes,
+                                 int64_t eachOperandMaxTileBits,
+                                 int64_t allOperandsMaxTileBits) {
+  int numLoops = op.getNumLoops();
+  assert(numLoops == vecTileSizes.size());
+  auto indexingMaps = op.getIndexingMapsArray();
+  auto operandTypes = op->getOperandTypes();
+  int numOperands = operandTypes.size();
+
+  // For each operand, we track how big the tile is going to be based on the
+  // dimensions that we have already accounted for. Here we initialize this to
+  // just elemBitWidth, having not yet accounted for any array dimension.
+  SmallVector<int> tileBits(numOperands);
+  for (int i : llvm::seq<int>(0, numOperands)) {
+    tileBits[i] = operandTypes[i].cast<ShapedType>().getElementTypeBitWidth();
+    if (tileBits[i] > eachOperandMaxTileBits) {
+      // eachOperandMaxTileBits is too small for this requirement to be
+      // satisfied. Early-returning. This might conceivably happen due to the
+      // tail-recursion in this function, where eachOperandMaxTileBits is
+      // halved.
+      return;
+    }
+  }
+
+  // For each loop, we must make sure that if it's the inner-most dimension that
+  // concerns an operand with a less-than-8-bit element type, we preserve the
+  // requirement that the inner-most tile dimension is >= 8 bit. For example, if
+  // the element type is `i2`, the minimum inner tile size is 8/2 = 4.
+  SmallVector<int64_t> minVecTileSizes(numLoops);
+  for (int loopNum : llvm::seq<int>(0, numLoops)) {
+    minVecTileSizes[loopNum] = 1;
+    for (int i : llvm::seq<int>(0, numOperands)) {
+      if (isInnerMostDimThatMapIsFunctionOf(indexingMaps[i], loopNum)) {
+        int64_t elemBitWidth =
+            operandTypes[i].cast<ShapedType>().getElementTypeBitWidth();
+        if (elemBitWidth < 8) {
+          minVecTileSizes[loopNum] =
+              std::max(minVecTileSizes[loopNum], 8 / elemBitWidth);
+        }
+      }
+    }
+  }
+
+  // For each loop starting from the inner-most, clamp vecTileSizes to honor
+  // `eachOperandMaxTileBits`, to the extent permitted by `minVecTileSizes`
+  // (which is a harder requirement).
+  for (int loopNum : llvm::reverse(llvm::seq<int>(0, numLoops))) {
+    // Skip 0 vecTileSizes, want to preserve them as 0.
+    if (vecTileSizes[loopNum] == 0) {
+      continue;
+    }
+    for (int i : llvm::seq<int>(0, numOperands)) {
+      // Check if this operand is concerned with this loop.
+      if (!indexingMaps[i].isFunctionOfDim(loopNum)) {
+        continue;
+      }
+      int64_t oldVal = vecTileSizes[loopNum];
+      int64_t maxVal = std::max(minVecTileSizes[loopNum],
+                                eachOperandMaxTileBits / tileBits[i]);
+      int64_t adjustedVal = std::min(oldVal, maxVal);
+      // If the old tile size was a power of two, make sure that we preserve
```

Comment:
If the old tile size was not Po2, but it is still adjusted here, wouldn't it still be good to keep it as a Po2?

If the tile size was not Po2 to begin with, then there may have been a specific reason for the non-Po2 size. In a case like this, if the tile size is getting clamped, then the specific tile size is already lost. I would think that it is better to at least clamp to a Po2 in this case. Alternatively, there was no reason for the non-Po2 tile size, and it is probably better off as Po2 anyway. WDYT?

Reply:
Oh, very nice catch! Yes, that makes sense. This results in a logic simplification: now the condition becomes just `if (adjustedVal != oldVal) {`. Thanks for looking into this level of detail.